// @hashgraph/cryptography/lib/primitive/aes.native.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CipherAlgorithm = void 0;
exports.createCipheriv = createCipheriv;
exports.createDecipheriv = createDecipheriv;
exports.messageDigest = messageDigest;
var _cryptoJs = _interopRequireDefault(require("crypto-js"));
var hex = _interopRequireWildcard(require("../encoding/hex.cjs"));
var utf8 = _interopRequireWildcard(require("../encoding/utf8.cjs"));
var _sparkMd = _interopRequireDefault(require("spark-md5"));
var _buffer = require("buffer");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const CipherAlgorithm = {
  Aes128Ctr: "AES-128-CTR",
  Aes128Cbc: "AES-128-CBC"
};

/**
 * @param {string} algorithm
 * @param {Uint8Array} key
 * @param {Uint8Array} iv
 * @param {Uint8Array} data
 * @returns {Promise<Uint8Array>}
 */
exports.CipherAlgorithm = CipherAlgorithm;
function createCipheriv(algorithm, key, iv, data) {
  let mode;
  switch (algorithm.toUpperCase()) {
    case CipherAlgorithm.Aes128Cbc:
      mode = _cryptoJs.default.mode.CBC;
      break;
    case CipherAlgorithm.Aes128Ctr:
      mode = _cryptoJs.default.mode.CTR;
      break;
    default:
      throw new Error("(BUG) non-exhaustive switch statement");
  }
  const data_ = _cryptoJs.default.enc.Hex.parse(hex.encode(data));
  const key_ = _cryptoJs.default.enc.Hex.parse(hex.encode(key.slice(0, 16)));
  const cfg_ = {
    iv: _cryptoJs.default.enc.Hex.parse(hex.encode(iv)),
    mode
  };
  return Promise.resolve(hex.decode(_cryptoJs.default.AES.encrypt(data_, key_, cfg_).toString(_cryptoJs.default.format.Hex)));
}

/**
 * @param {string} algorithm
 * @param {Uint8Array} key
 * @param {Uint8Array} iv
 * @param {Uint8Array} data
 * @returns {Promise<Uint8Array>}
 */
function createDecipheriv(algorithm, key, iv, data) {
  let mode;
  switch (algorithm) {
    case CipherAlgorithm.Aes128Cbc:
      mode = _cryptoJs.default.mode.CBC;
      break;
    case CipherAlgorithm.Aes128Ctr:
      mode = _cryptoJs.default.mode.CTR;
      break;
    default:
      throw new Error("(BUG) non-exhaustive switch statement");
  }
  const key_ = _cryptoJs.default.enc.Hex.parse(hex.encode(key.slice(0, 16)));
  const iv_ = _cryptoJs.default.enc.Hex.parse(hex.encode(iv));
  const params = _cryptoJs.default.lib.CipherParams.create({
    ciphertext: _cryptoJs.default.enc.Hex.parse(hex.encode(data)),
    iv: iv_,
    key: key_,
    algorithm: _cryptoJs.default.algo.AES,
    blockSize: 4
  });
  return Promise.resolve(hex.decode(_cryptoJs.default.AES.decrypt(params, key_, {
    iv: iv_,
    mode
  }).toString(_cryptoJs.default.enc.Hex)));
}

/**
 * @param {string} passphrase
 * @param {string} iv
 * @returns {Promise<Uint8Array>}
 */
async function messageDigest(passphrase, iv) {
  const pass = utf8.encode(passphrase);
  const sliced = hex.decode(iv).slice(0, 8);
  const result = _sparkMd.default.ArrayBuffer.hash(_buffer.Buffer.concat([_buffer.Buffer.from(pass), _buffer.Buffer.from(sliced)]));
  return Promise.resolve(hex.decode(result));
}
