// @hashgraph/sdk/lib/network/NetworkVersionInfo.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _SemanticVersion = _interopRequireDefault(require("./SemanticVersion.cjs"));
var HashgraphProto = _interopRequireWildcard(require("@hashgraph/proto"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
/*-
 * ‌
 * Hedera JavaScript SDK
 * ​
 * Copyright (C) 2020 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

/**
 * Response when the client sends the node CryptoGetVersionInfoQuery.
 */
class NetworkVersionInfo {
  /**
   * @private
   * @param {object} props
   * @param {SemanticVersion} props.protobufVersion
   * @param {SemanticVersion} props.servicesVesion
   */
  constructor(props) {
    /**
     * The account ID for which this information applies.
     *
     * @readonly
     */
    this.protobufVersion = props.protobufVersion;

    /**
     * The account ID for which this information applies.
     *
     * @readonly
     */
    this.servicesVesion = props.servicesVesion;
    Object.freeze(this);
  }

  /**
   * @internal
   * @param {HashgraphProto.proto.INetworkGetVersionInfoResponse} info
   * @returns {NetworkVersionInfo}
   */
  static _fromProtobuf(info) {
    return new NetworkVersionInfo({
      protobufVersion: _SemanticVersion.default._fromProtobuf( /** @type {HashgraphProto.proto.ISemanticVersion} */
      info.hapiProtoVersion),
      servicesVesion: _SemanticVersion.default._fromProtobuf( /** @type {HashgraphProto.proto.ISemanticVersion} */
      info.hederaServicesVersion)
    });
  }

  /**
   * @internal
   * @returns {HashgraphProto.proto.INetworkGetVersionInfoResponse}
   */
  _toProtobuf() {
    return {
      hapiProtoVersion: this.protobufVersion._toProtobuf(),
      hederaServicesVersion: this.servicesVesion._toProtobuf()
    };
  }

  /**
   * @param {Uint8Array} bytes
   * @returns {NetworkVersionInfo}
   */
  static fromBytes(bytes) {
    return NetworkVersionInfo._fromProtobuf(HashgraphProto.proto.NetworkGetVersionInfoResponse.decode(bytes));
  }

  /**
   * @returns {Uint8Array}
   */
  toBytes() {
    return HashgraphProto.proto.NetworkGetVersionInfoResponse.encode(this._toProtobuf()).finish();
  }
}
exports.default = NetworkVersionInfo;
