// @hashgraph/sdk/lib/FeeSchedules.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var HashgraphProto = _interopRequireWildcard(require("@hashgraph/proto"));
var _FeeSchedule = _interopRequireDefault(require("./FeeSchedule.cjs"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*-
 * ‌
 * Hedera JavaScript SDK
 * ​
 * Copyright (C) 2020 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

class FeeSchedules {
  /**
   * @param {object} [props]
   * @param {FeeSchedule} [props.currentFeeSchedule]
   * @param {FeeSchedule} [props.nextFeeSchedule]
   */
  constructor(props = {}) {
    /*
     * Contains current Fee Schedule
     *
     * @type {FeeSchedule}
     */
    this.current = props.currentFeeSchedule;

    /*
     * Contains next Fee Schedule
     *
     * @type {FeeSchedule}
     */
    this.next = props.nextFeeSchedule;
  }

  /**
   * @param {Uint8Array} bytes
   * @returns {FeeSchedules}
   */
  static fromBytes(bytes) {
    return FeeSchedules._fromProtobuf(HashgraphProto.proto.CurrentAndNextFeeSchedule.decode(bytes));
  }

  /**
   * @internal
   * @param {HashgraphProto.proto.ICurrentAndNextFeeSchedule} feeSchedules
   * @returns {FeeSchedules}
   */
  static _fromProtobuf(feeSchedules) {
    return new FeeSchedules({
      currentFeeSchedule: feeSchedules.currentFeeSchedule != null ? _FeeSchedule.default._fromProtobuf(feeSchedules.currentFeeSchedule) : undefined,
      nextFeeSchedule: feeSchedules.nextFeeSchedule != null ? _FeeSchedule.default._fromProtobuf(feeSchedules.nextFeeSchedule) : undefined
    });
  }

  /**
   * @internal
   * @returns {HashgraphProto.proto.ICurrentAndNextFeeSchedule}
   */
  _toProtobuf() {
    return {
      currentFeeSchedule: this.current != null ? this.current._toProtobuf() : undefined,
      nextFeeSchedule: this.next != null ? this.next._toProtobuf() : undefined
    };
  }

  /**
   * @returns {Uint8Array}
   */
  toBytes() {
    return HashgraphProto.proto.CurrentAndNextFeeSchedule.encode(this._toProtobuf()).finish();
  }
}
exports.default = FeeSchedules;
